## Search

1. Data is a finite sequence of codas.
2. A coda is a pair of data. 

This defines and countably infinite universe of data.  We can consider cutting these off at some maximum depth or width.  This could be consistently implemented by defining the "outer crust" to be atomic or by mapping them all to empty data.  All sorts of interesting questions arise.

In [1]:
import itertools
import Search

def f(depth,width): 
    U = Search.alldata(depth,width)
    print(depth,width,'->',len(U),len(set(U)))

In [2]:
f(5,1)

5 1 -> 458330 458330


### Data sample sizes width (horizontal) vs depth (vertical)

| width:   |    1    |    2   |    3   |    4    |    5     |
|---------:|--------:|-------:|-------:|--------:|---------:|
| depth  1      |    2    |   3    |   4     |    5     |  6  |
| depth  2      |    5    |  91    | 4369   | 406901  |    |
| depth  3      |   26    | 68583243     |        |         |   |
| depth  4      |  677    |        |        |         |    |
| depth  5      | 458330  |        |        |         |    |

 

In [3]:
n = 0 
for d in Search.alldata(2,2): 
    n += 1 
    print(n,'<-',repr(d),len(d),d.depth())

1 <-  0 0
2 <- (:) 1 1
3 <- (:(:)) 1 2
4 <- (:(:)(:)) 1 2
5 <- ((:):) 1 2
6 <- ((:):(:)) 1 2
7 <- ((:):(:)(:)) 1 2
8 <- ((:)(:):) 1 2
9 <- ((:)(:):(:)) 1 2
10 <- ((:)(:):(:)(:)) 1 2
11 <- (:)(:) 2 1
12 <- (:)(:(:)) 2 2
13 <- (:)(:(:)(:)) 2 2
14 <- (:)((:):) 2 2
15 <- (:)((:):(:)) 2 2
16 <- (:)((:):(:)(:)) 2 2
17 <- (:)((:)(:):) 2 2
18 <- (:)((:)(:):(:)) 2 2
19 <- (:)((:)(:):(:)(:)) 2 2
20 <- (:(:))(:) 2 2
21 <- (:(:))(:(:)) 2 2
22 <- (:(:))(:(:)(:)) 2 2
23 <- (:(:))((:):) 2 2
24 <- (:(:))((:):(:)) 2 2
25 <- (:(:))((:):(:)(:)) 2 2
26 <- (:(:))((:)(:):) 2 2
27 <- (:(:))((:)(:):(:)) 2 2
28 <- (:(:))((:)(:):(:)(:)) 2 2
29 <- (:(:)(:))(:) 2 2
30 <- (:(:)(:))(:(:)) 2 2
31 <- (:(:)(:))(:(:)(:)) 2 2
32 <- (:(:)(:))((:):) 2 2
33 <- (:(:)(:))((:):(:)) 2 2
34 <- (:(:)(:))((:):(:)(:)) 2 2
35 <- (:(:)(:))((:)(:):) 2 2
36 <- (:(:)(:))((:)(:):(:)) 2 2
37 <- (:(:)(:))((:)(:):(:)(:)) 2 2
38 <- ((:):)(:) 2 2
39 <- ((:):)(:(:)) 2 2
40 <- ((:):)(:(:)(:)) 2 2
41 <- ((:):)((:):) 2 2
42 <- ((:):)((:):(:)) 2 

In [4]:
datas = [d for d in Search.alldata(2,3)]
len(datas)

4369

In [5]:
I = Search.Idempotents(datas,datas)
I.search()
print(I)

1 ... {}
2 ... {'never true': 1}
3 ... {'never true': 2}
4 ... {'never true': 2, 'never false': 1}
5 ... {'never true': 2, 'never false': 2}
6 ... {'never true': 2, 'never false': 3}
7 ... {'never true': 3, 'never false': 3}
8 ... {'never true': 4, 'never false': 3}
9 ... {'never true': 4, 'never false': 4}
10 ... {'never true': 4, 'never false': 5}
11 ... {'never true': 4, 'never false': 6}
12 ... {'never true': 4, 'never false': 7}
13 ... {'never true': 4, 'never false': 8}
14 ... {'never true': 4, 'never false': 9}
15 ... {'never true': 4, 'never false': 10}
16 ... {'never true': 4, 'never false': 11}
17 ... {'never true': 4, 'never false': 12}
18 ... {'never true': 4, 'never false': 13}
19 ... {'never true': 5, 'never false': 13}
20 ... {'never true': 6, 'never false': 13}
21 ... {'never true': 7, 'never false': 13}
22 ... {'never true': 8, 'never false': 13}
23 ... {'never true': 9, 'never false': 13}
24 ... {'never true': 10, 'never false': 13}
25 ... {'never true': 11, 'never fa

KeyboardInterrupt: 

In [ ]:
for d,res in I._result.items():
    if res.undecided(): print(d,'--->',res)